Update language reference for nullable token and related topics (#17730)
dotnet/docs: docs/csharp/language-reference/builtin-types/nullable-reference-types.md

---
title: "Nullable reference types - C# reference"
description: Learn about C# nullable reference types and how to use them
ms.date: 04/06/2020
---
# Nullable reference types (C# reference)

> [!NOTE]
> This article covers nullable reference types. You can also declare [nullable value types](nullable-value-types.md).

Nullable reference types are available beginning with C# 8.0, in code that has opted in to a *nullable aware context*. Nullable reference types, the null static analysis warnings, and the null forgiveness operator are optional language features. All are turned off by default. A *nullable context* is controlled at the project level using build settings, or in code using pragmas.

 In a nullable aware context:

- A reference type `T` must be initialized to non-null, and may never be assigned a value that may be `null`.

> - A reference type `T?` may be initialized or assigned to `null`, but is required to be checked against `null` before de-referencing.

Author: I read this and proof-read it to make sure I distinguished "variable of ...type" vs. "type" and I still missed these. Great catch.

- A variable `m` of type `T?` is considered to be non-null when you apply the null-forgiving operator, as in `m!`.

> The distinctions between a non-nullable reference type `T` and a nullable reference type `T?` are performed by the compiler's interpretation of the preceding rules. A variable of type `T` and a variable of type `T?` are represented by the same .NET type. The following example declares a non-nullable string, a nullable string, and uses the null-forgiving operator to assign a value to a non-nullable string:

Reviewer: I thought the original wording that was replaced by "the same .NET type" was clearer.

```suggestion
The distinctions between a non-nullable reference type `T` and a nullable reference type `T?` are enforced by the compiler's interpretation of the preceding rules. A variable of type `T` and a variable of type `T?` are represented by the same type in IL and at run time. The following example declares a non-nullable string and a nullable string, and then uses the null-forgiving operator to assign a value to a non-nullable string:
```

Author: To avoid the conflicts, I added those suggestions by hand.

:::code language="csharp" source="snippets/NullableReferenceTypes.cs" id="SnippetCoreSyntax":::

> The variable `notNull` and `nullable` are both represented by the <xref:System.String> type. Because the non-nullable and nullable types are both stored as the same type, there are several locations where using a nullable reference type isn't allowed. In general, a nullable reference type can't be used as a base class or implemented interface. A nullable reference type can't be used in any object creation or type testing expression. A nullable reference type can't be the type of a member access expression. The following examples show these constructs:

Reviewer 2: > Because the non-nullable and nullable types are both stored as the same type

That reads not clear. Can that whole sentence be removed, and then just list the locations where nullable reference types aren't allowed?

```suggestion
The variables `notNull` and `nullable` are both represented by the <xref:System.String> type. Because the non-nullable and nullable types are both stored as the same type, there are several locations where using a nullable reference type isn't allowed. In general, a nullable reference type can't be used as a base class or implemented interface. A nullable reference type can't be used in any object creation or type testing expression. A nullable reference type can't be the type of a member access expression. The following examples show these constructs:
```

Author: I'll take this suggestion, and consider making a list.

> The variables `notNull` and `nullable` are both represented by the <xref:System.String> type. Because the non-nullable and nullable types are both stored as the same type, there are several locations where using a nullable reference type isn't allowed. In general, a nullable reference type can't be used as a base class or implemented interface. A nullable reference type can't be used in any object creation or type testing expression. A nullable reference type can't be the type of a member access expression. The following examples show these constructs:

Reviewer: >The following examples show these constructs:

I don't see the implemented interface example.

I think this might be easier to follow if it were a bulleted list with the description of each example preceding the code for that scenario.

```csharp
public MyClass : System.Object? // not allowed
{
}

var nullEmpty = System.String?.Empty; // Not allowed
var maybeObject = new object?(); // Not allowed
try
{
    if (thing is string? nullableString) // not allowed
        Console.WriteLine(nullableString);
} catch (Exception? e) // Not Allowed
{
    Console.WriteLine("error");
}
```

## Nullable references and static analysis

> The examples in the previous section illustrate the nature of nullable reference types: Nullable reference types aren't new class types, but rather annotations on existing reference types. The compiler uses those annotations to help you find potential null reference errors in your code. There's no runtime difference between a non-nullable reference type and a nullable reference type. The compiler doesn't add any runtime checking for non-nullable reference types. The benefits are in the compile-time analysis. The compiler generates warnings that help you find and fix potential null errors in your code. You declare your intent, and the compiler warns you when your code violates that intent.

Reviewer 2: ```suggestion
The examples in the previous section illustrate the nature of nullable reference types. Nullable reference types aren't new class types, but rather annotations on existing reference types. The compiler uses those annotations to help you find potential null reference errors in your code. There's no runtime difference between a non-nullable reference type and a nullable reference type. The compiler doesn't add any runtime checking for non-nullable reference types. The benefits are in the compile-time analysis. The compiler generates warnings that help you find and fix potential null errors in your code. You declare your intent, and the compiler warns you when your code violates that intent.
```

The following snippet shows where the compiler emits warnings when using this class:

:::code language="csharp" source="snippets/NullableReferenceTypes.cs" id="SnippetLocalWarnings":::

The preceding examples demonstrate the compiler's static analysis to determine the null state of reference variables. The compiler applies language rules for null checks and assignments to inform its analysis.  The compiler can't make assumptions about the semantics of methods or properties. If you call methods that perform null checks, the compiler can't know those methods affect a variable's null state. There are a number of attributes you can add to your APIs to inform the compiler about the semantics of arguments and return values. These attributes have been applied to many common APIs in the .NET Core libraries. For example, <xref:System.String.IsNullOrEmpty%2A> has been updated, and the compiler correctly interprets that method as a null check. For more information about the attributes that apply to null state static analysis, see the article on [Nullable attributes](../../nullable-attributes.md).

## Setting the nullable context

> There are two ways to control the nullable context. At the project level, you can add the `<Nullable>Enable</Nullable>` project setting. In a single C# source file, you can add the `#nullable enable` pragma to enable the nullable context. See the article on [setting a nullable strategy](../../nullable-attributes.md).

Reviewer 2: ```suggestion
There are two ways to control the nullable context. At the project level, you can add the `<Nullable>enable</Nullable>` project setting. In a single C# source file, you can add the `#nullable enable` pragma to enable the nullable context. See the article on [setting a nullable strategy](../../nullable-attributes.md).
```
